Test/QuickCheck/Monadic.hs

```diff
@@ -273,5 +291,65 @@ runSTGen f = do
   return (runST (eval f))
 #endif
 
+-- Exceptions
+
+
+#ifndef NO_EXCEPTIONS
+
+-- | Evaluate the value to WHNF and check if it raises an exception.
+assertException ::
+     (E.Exception exc)
+  => (exc -> Bool) -- ^ Return True if that is the exception that was expected
+  -> a -- ^ Value that should throw an exception, when evaluated
+  -> Property
+assertException isExc action = assertExceptionIO isExc (return action)
+
+
+-- | Make sure that an exception is thrown during an IO action. The result is
+-- evaluated to WHNF.
+assertExceptionIO ::
+     (E.Exception exc)
+  => (exc -> Bool) -- ^ Return True if that is the exception that was expected
+  -> IO a -- ^ An action that should throw the exception
+  -> Property
+assertExceptionIO isExc action =
+  monadicIO $ do
+    hasFailed <-
+      run
+        (E.catch
+           (do res <- action
+               res `seq` return False)
+           (return . isExc))
+    assert hasFailed
+
+#ifndef NO_DEEPSEQ
+
+-- | Evaluate the value to NF and check if it raises an exception.
+assertDeepException ::
+     (NFData a, E.Exception exc)
+  => (exc -> Bool) -- ^ Return True if that is the exception that was expected
+  -> a -- ^ Value that should throw an exception, when fully evaluated
+  -> Property
+assertDeepException isExc action = assertDeepExceptionIO isExc (return action)
+
+-- | Make sure that an exception is thrown during an IO action. The result is
+-- evaluated to WHNF.
+assertDeepExceptionIO ::
+     (NFData a, E.Exception exc)
+  => (exc -> Bool) -- ^ Return True if that is the exception that was expected
+  -> IO a -- ^ An action that should throw the exception
+  -> Property
+assertDeepExceptionIO isExc action =
```

Comment:
Same as above?

Reply:
There is definitely no point in this code duplication. 
It's been some time since I wrote those functions, so I am not sure why I went with copy/paste solution :smile:

Reply:
I did think it was a bit out of character! But it's a long time ago ;)